### PR TITLE
microbit: add basic support for LED matrix

### PR DESCRIPTION
**Pull request by deadprogram**
Add very basic functions for LED matrix and correct button pin mapping for BBC micro:bit

### CODE REVIEW

**Comment by deadprogram** [comment]
This code works:

```go
package main

import (
	"machine"
	"time"
)

func main() {
	machine.InitLEDMatrix()

	left := machine.GPIO{machine.BUTTONA}
	left.Configure(machine.GPIOConfig{Mode: machine.GPIO_INPUT})

	right := machine.GPIO{machine.BUTTONB}
	right.Configure(machine.GPIOConfig{Mode: machine.GPIO_INPUT})

	var (
		x uint8 = 2
		y uint8 = 2
	)

	for {
		machine.SetLEDMatrix(x, y, false)
		if !left.Get() {
			switch {
			case x > 0:
				x--
			case x == 0:
				if y > 0 {
					x = 4
					y--
				}
			}
		}

		if !right.Get() {
			switch {
			case x < 4:
				x++
			case x == 4:
				if y < 4 {
					x = 0
					y++
				}
			}
		}

		machine.SetLEDMatrix(x, y, true)
		time.Sleep(time.Millisecond * 100)
	}
}
```

Not sure where I should put it... what do you think?

**Comment by aykevl** [comment]
I'm actually wondering whether this repo is the right place for such drivers. Maybe this one can go in because it is highly board-specific, but it will prevent the re-use of such a driver for other boards.

Also, properly implementing the LED matrix (with brightness levels and all) is pretty difficult. IIRC it requires very tight code using the TIMER peripheral to switch LEDs on/off at the right time from within an interrupt.

For the future, I'm imagining there could be a new "drivers" repository which can for example contain drivers for specific (external) chips, like the compass and accelerometer on the micro:bit but maybe also drivers for bitbanged protocols like DHT22 and WS2812. In such a repository, every chip can be its own package, like "\<repo>/driver/mag3110" and "\<repo>/drivers/mma8653fc". This is especially convenient for I2C using chips because they can simply take a `machine.I2C` type and work with that:

```go
machine.I2C0.Configure(machine.I2CConfig{...})
mag := mag3110.New(machine.I2C0)
for {
    x, y, z := mag.ReadMagnetic() // returns int16 values
    println("magnetic components:", x, y, z)
    time.Sleep(time.Second)
}
```

Putting these drivers in a new repository will make updating easier as the compiler doesn't have to be updated every time to add a new driver. (I realize the same could be true for the machine package...)

I wanted to register github.com/tinygo to collect these repositories but it appears to be taken... You haven't by any chance registered this one in advance? ;)

**Comment by deadprogram** [comment]
I was thinking the same thing about device drivers and where they should go.

Regarding the tinygo organization, unfortunately it was not I who registered it. Also it does not seem like it is active... cause now would be a great time to have that organization, with the possibility of having another repo.

**Comment by deadprogram** [comment]
Regarding the LED matrix code I submitted, it is just a really basic way to get some visual feedback from the LEDs, not a full "frame render" like you describe which is the "real" solution. Probably it will do for now as far as it goes since highly board-specific functionality. Eventually can probably go in `tinygo-machines` or where ever that ends up.

**Comment by aykevl** [comment]
I just tried the example code and it's pretty cool :)
It would be nice to have example code of the LED matrix somewhere. I put mine in "examples/microbitmatrix" but I'm not really happy with the name. If you have a better suggestion please tell.

Regarding the function `SetLEDMatrix`:
It's a bit of a leaky abstraction because you have to clear a given LED before you can set another. I'm not sure what a good solution would be and maybe we'd just have to leave it as is. One solution could be dropping the 'on' parameter and disabling all LEDs before setting a new one.
In any case, I think it would be good to add a comment to the function indicating that whenever you set a LED, you have to clear it again before setting another LED.

File: src/machine/board_microbit.go [hunk 1]
```diff
@@ -1,15 +1,20 @@
...
 const HasLowFrequencyCrystal = false
 
 // Buttons on the micro:bit (A and B)
 const (
-	BUTTON  = BUTTON1
-	BUTTON1 = 5
-	BUTTON2 = 11
+	BUTTON  = BUTTONA
+	BUTTONA = 17
+	BUTTONB = 26
```
Comment (aykevl) [inline src/machine/board_microbit.go]: Weird. I checked the pinout for the buttons and they were supposed to be on 5 and 11. Not sure what's wrong with the docs or why they have to be different numbers.

Comment (deadprogram) [reply src/machine/board_microbit.go]: The numbering of the pins on the connector is not the same as the pins on the nrf51. That is why the numbers were off. You can find the mappings here:
https://github.com/lancaster-university/microbit-dal/blob/master/inc/drivers/MicroBitPin.h#L41

**Comment by deadprogram** [comment]
Updated this PR to reflect the real way that the LED matrix function operate currently.

**Comment by deadprogram** [comment]
My current example is located here: https://github.com/hybridgroup/tinygo-zoo/blob/master/pixelbuttons/pixelbuttons.go

File: src/machine/board_microbit.go [hunk 2]
```diff
@@ -30,3 +35,60 @@ const (
...
+		nrf.GPIO.DIRSET = 1 << uint8(i)
+	}
+	ClearLEDMatrix()
+}
+
+// ClearLEDMatrix clears the entire LED matrix.
+func ClearLEDMatrix() {
+	for i := LED_COL_1; i <= LED_COL_9; i++ {
+		nrf.GPIO.OUTSET = 1 << uint8(i)
+	}
```
Comment (aykevl) [inline src/machine/board_microbit.go]: This produces smaller code (32 bytes less) and is probably faster as well:

```go
// ClearLEDMatrix clears the entire LED matrix.
func ClearLEDMatrix() {
        set := 0
        for i := LED_COL_1; i <= LED_COL_9; i++ {
                set |= 1 << uint8(i)
        }
        nrf.GPIO.OUTSET = nrf.RegValue(set)
        nrf.GPIO.OUTCLR = (1 << LED_ROW_1) | (1 << LED_ROW_2) | (1 << LED_ROW_3)
}
```

The reason, again, is that direct register access is expensive and with this change the whole `for` loop can be optimized away.

Comment (deadprogram) [reply src/machine/board_microbit.go]: Good to know. Change pushed.

**Comment by aykevl** [comment]
Merged in 6b89059f5f33e120e5af2e82e3741d7e08d1c5cd. Thank you!

**Comment by aykevl** [comment]
I force-pushed the last commit you did, I hope nobody noticed...
It is now in c516a11004d600e4b876e3d6589206b878bcd8a6.

**Pull request closed without merging**